rcli/call.py

```diff
@@ -0,0 +1,405 @@
+)
+
+_JSON_TYPES = (
+    dict,
+    typing.Mapping,
+    typing.MutableMapping
+)
+
+NoneType = type(None)
+
+
+def call(func, args):
+    """Call the function with args normalized and cast to the correct types.
+
+    Args:
+        func: The function to call.
+        args: The arguments parsed by docopt.
+
+    Returns:
+        The return value of func.
+    """
+    norm_args = _normalize(func, args)
+    hints = _get_type_hints(func)
+    f_args = {k: _cast(hints[k], v) for k, v in six.iteritems(norm_args)}
+    return func(**f_args)
+
+
+def get_callable(subcommand):
+    """Return a callable object from the subcommand.
+
+    Args:
+        subcommand: A object loaded from an entry point. May be a module,
+            class, or function.
+
+    Returns:
+        The callable entry point for the subcommand. If the subcommand is a
+        function, it will be returned unchanged. If the subcommand is a module
+        or a class, an instance of the command class will be returned.
+
+    Raises:
+        AssertionError: Raised when a module entry point does not have a
+            callable class named Command.
+    """
+    _LOGGER.debug(
+        'Creating callable from subcommand "%s".', subcommand.__name__)
+    if isinstance(subcommand, types.ModuleType):
+        _LOGGER.debug('Subcommand is a module.')
+        assert hasattr(subcommand, 'Command'), (
+            'Module subcommand must have callable "Command" class definition.')
+        subcommand = subcommand.Command
+    if any(isinstance(subcommand, t) for t in six.class_types):
+        return subcommand()
+    return subcommand
+
+
+def _normalize(func, cli_args):
+    """Alter the docopt args to be valid python names for func.
+
+    Returns a dictionary based on cli_args that uses normalized keys. Keys are
+    normalized to valid python names.
+
+    Args:
+        func: The function being called to run the command.
+        cli_args: The parsed results of docopt for the given command.
+
+    Returns:
+        A dictionary containing normalized keys from the CLI arguments. If the
+        CLI arguments contains values that the function will not accept, those
+        keys will not be returned.
+    """
+    assert hasattr(func, '__call__'), (
+        'Cannot normalize parameters of a non-callable.')
+    is_func = isinstance(func, types.FunctionType)
+    params = _get_signature(func if is_func else func.__call__)
+    args = {}
+    multi_args = set()
+    for k, v in six.iteritems(cli_args):
+        nk = re.sub(r'\W|^(?=\d)', '_', k).strip('_').lower()
+        if nk in params:
+            if nk not in args or args[nk] is None:
+                args[nk] = v
+            elif nk in multi_args and v is not None:
+                args[nk].append(v)
+            elif v is not None:
+                multi_args.add(nk)
+                args[nk] = [args[nk], v]
+    _LOGGER.debug('Normalized "%s" to "%s".', cli_args, args)
+    return args
+
+
+def _get_signature(func):
```

Comment:
This function was moved from dispatcher.py. Formatting was changed slightly, but the code was not altered.